docs/specifications/tests/Zone-TP/zone11.md

```diff
@@ -0,0 +1,262 @@
+   ("Name Server IP").
+
+4. For each name server in *Name Server IP* do:
+
+   1. Send *TXT Query* to the name server and collect the [DNS Response].
+
+   2. Go to the next name server IP address if at least one of the following
+      criteria is met:
+
+      1. There is no DNS response.
+      2. [RCODE Name] of the response is not "NoError".
+      3. The AA flag is not set in the response.
+      4. There is no TXT resource record in the response.
+
+   3. If the name server responds with at least one TXT record, then, for each
+      TXT record:
+
+      1. [Concatenate][concatenate] all strings within the TXT record’s data
+         together.
+      2. If the resulting string starts with `v=spf`, then add it to the
+         *SPF-Policies* set.
+
+   4. Go to the next name server.
+
+5. If the *SPF-Policies* set is empty, then output *[Z11_SPF_NOT_CONFIGURED]*
+   and terminate the test.
+
+6. For each string in *SPF-Policies* set, if the string is equal to `v=spf1` or
+   starts with `v=spf1` followed by a space, then add it to the *SPFv1-Policies*
+   set.
+
+7. If the *SPFv1-Policies* set is empty, then for each string in the
+   *SPF-Policies* set, output *[Z11_SPF_WRONG_VERSION]* with the version number
+   extracted from the string.
+
+8. If the *SPFv1-Policies* set contains more than one element, then output
+   *[Z11_SPF1_MULTIPLE_RECORDS]* and terminate the test. The next steps assume
+   there is exactly one string in the *SPFv1-Policies* set ("SPF policy").
+
+9. If the *SPF Policy* does not [pass the syntax check][passing the syntax
+   check] for SPF version 1 records, then output *[Z11_SPF1_SYNTAX_ERROR]* and
+   terminate the test.
+
+10. If the *SPF Policy* contains more than one `redirect` modifier, or more than
+    one `exp` modifier, then output *[Z11_SPF1_MULTIPLE_SAME_MODIFIER]*.
+
+11. If the *SPF Policy* contains exactly one `redirect` modifier anywhere except
+    at the end of the policy text, then output *[Z11_SPF1_REDIRECT_NOT_AT_END]*.
+
+12. If the *SPF Policy* contains both a `redirect` modifier and an `all`
+    mechanism, then output *[Z11_SPF1_REDIRECT_AND_ALL]*.
+
+13. If the *SPF Policy* contains at least one `ptr` mechanism, then output
+    *[Z11_SPF1_HAS_PTR]*.
+
+14. Count ("DNS Lookup count") the number of terms in the *SPF Policy* that
+    are either:
+
+   1. A mechanism of type `a`, `mx`, `ptr`, `exists`;
+
+   2. Or a modifier of type `redirect`.
+
+15. If the `%{p}` [macro] appears as a domain-spec for an [SPF term] in the
+    *SPF Policy*, then:
+
+   1. Output *[Z11_SPF1_HAS_P_MACRO]*.
+
+   2. If the same policy contains no `ptr` mechanism, then increment the *DNS
+   Lookup count*.
+
+16. If the *DNS Lookup count* is greater than 10, then output
+    *[Z11_SPF1_TOO_COMPLEX]*.
+
+17. If the *DNS Lookup count* is less than or equal to 10 and the *SPF Policy*
+    contains a `redirect` modifier or at least one `include` mechanism, then
+    output *[Z11_SPF1_RECURSIVE]*.
+
```

Comment:
I’ll add `spf_policy` as an argument to all of these messages.

Reply:
You have renumbered. My comment was about returning the name server where the SPF record with problem was found.

Reply:
Okay, that makes sense too. I think it would be useful to have both the erroneous SPF policy and the nameserver that returns it.

As for the length of the SPF policies, let’s just start with including them in their entirety in the messages. We can evaluate later whether it turns out to be impractical. I’ll need to research the average length of an SPF policy.